plugins/vpp/srplugin/descriptor/localsid.go

```diff
@@ -0,0 +1,282 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package descriptor
+
+import (
+	"net"
+	"reflect"
+	"strings"
+
+	"github.com/ligato/cn-infra/logging"
+	interfaces "github.com/ligato/vpp-agent/api/models/vpp/interfaces"
+	srv6 "github.com/ligato/vpp-agent/api/models/vpp/srv6"
+	scheduler "github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	"github.com/ligato/vpp-agent/plugins/vpp/srplugin/descriptor/adapter"
+	"github.com/ligato/vpp-agent/plugins/vpp/srplugin/vppcalls"
+	"github.com/pkg/errors"
+)
+
+const (
+	// LocalSIDDescriptorName is the name of the descriptor for VPP LocalSIDs
+	LocalSIDDescriptorName = "vpp-sr-localsid"
+
+	// dependency labels
+	localsidOutgoingInterfaceDep = "sr-localsid-outgoing-interface-exists"
+	localsidIncomingInterfaceDep = "sr-localsid-incoming-interface-exists"
+)
+
+// LocalSIDDescriptor teaches KVScheduler how to configure VPP LocalSIDs.
+type LocalSIDDescriptor struct {
+	// dependencies
+	log       logging.Logger
+	srHandler vppcalls.SRv6VppAPI
+}
+
+// NewLocalSIDDescriptor creates a new instance of the LocalSID descriptor.
+func NewLocalSIDDescriptor(srHandler vppcalls.SRv6VppAPI, log logging.PluginLogger) *LocalSIDDescriptor {
+	return &LocalSIDDescriptor{
+		log:       log.NewLogger("localsid-descriptor"),
+		srHandler: srHandler,
+	}
+}
+
+// GetDescriptor returns descriptor suitable for registration (via adapter) with
+// the KVScheduler.
+func (d *LocalSIDDescriptor) GetDescriptor() *adapter.LocalSIDDescriptor {
+	return &adapter.LocalSIDDescriptor{
+		Name:               LocalSIDDescriptorName,
+		NBKeyPrefix:        srv6.ModelLocalSID.KeyPrefix(),
+		ValueTypeName:      srv6.ModelLocalSID.ProtoName(),
+		KeySelector:        srv6.ModelLocalSID.IsKeyValid,
+		KeyLabel:           srv6.ModelLocalSID.StripKeyPrefix,
+		ValueComparator:    d.EquivalentLocalSIDs,
+		Validate:           d.Validate,
+		Create:             d.Create,
+		Delete:             d.Delete,
+		UpdateWithRecreate: d.UpdateWithRecreate,
+		Dependencies:       d.Dependencies,
+	}
+}
+
+// EquivalentLocalSIDs determines whether 2 localSIDs are logically equal. This comparison takes into consideration also
+// semantics that couldn't be modeled into proto models (i.e. SID is IPv6 address and not only string)
+func (d *LocalSIDDescriptor) EquivalentLocalSIDs(key string, oldLocalSID, newLocalSID *srv6.LocalSID) bool {
+	if oldLocalSID == nil || newLocalSID == nil {
+		return oldLocalSID == newLocalSID
```

Comment:
nil checks are not needed - if the previous or the new value is nil, the operation to execute will be Create or Delete.  ValueComparator is only used for Update, therefore it will not be called in those cases.

Reply:
fixed in https://github.com/ligato/vpp-agent/pull/1174/commits/63aa9e15d7b692c62577c288db87eac9c802974f